File: app/src/main/java/wannabit/io/cosmostaion/activities/AppLockActivity.java
package wannabit.io.cosmostaion.activities;

import android.annotation.TargetApi;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.core.content.ContextCompat;
import androidx.core.hardware.fingerprint.FingerprintManagerCompat;
import androidx.core.os.CancellationSignal;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;

import java.util.ArrayList;

import wannabit.io.cosmostaion.R;
import wannabit.io.cosmostaion.base.BaseActivity;
import wannabit.io.cosmostaion.fragment.AlphabetKeyBoardFragment;
import wannabit.io.cosmostaion.fragment.KeyboardFragment;
import wannabit.io.cosmostaion.fragment.NumberKeyBoardFragment;
import wannabit.io.cosmostaion.task.TaskListener;
import wannabit.io.cosmostaion.task.TaskResult;
import wannabit.io.cosmostaion.task.UserTask.CheckPasswordTask;
import wannabit.io.cosmostaion.utils.KeyboardListener;
import wannabit.io.cosmostaion.utils.WUtil;
import wannabit.io.cosmostaion.widget.StopViewPager;

public class AppLockActivity extends BaseActivity implements KeyboardListener, TaskListener {

    private LinearLayout mLayerContents;
    private ImageView mFingerImage;
    private TextView mUnlockMsg;
    private ImageView[] mIvCircle = new ImageView[5];

    private StopViewPager mViewPager;
    private KeyboardPagerAdapter mAdapter;
    private String mUserInput = "";

    private FingerprintManagerCompat mFingerprintManagerCompat;
    private CancellationSignal mCancellationSignal;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_applock);
        mLayerContents = findViewById(R.id.layer_contents);
        mFingerImage = findViewById(R.id.img_fingerprint);
        mUnlockMsg = findViewById(R.id.tv_unlock_msg);
        mViewPager = findViewById(R.id.pager_keyboard);
        mNeedLeaveTime = false;

        for (int i = 0; i < mIvCircle.length; i++) {
            mIvCircle[i] = findViewById(getResources().getIdentifier("img_circle" + i, "id", getPackageName()));
        }

        mViewPager.setOffscreenPageLimit(2);
        mAdapter = new KeyboardPagerAdapter(getSupportFragmentManager());
        mViewPager.setAdapter(mAdapter);


    }

    @Override
    protected void onResume() {
        super.onResume();
        mFingerImage.setColorFilter(ContextCompat.getColor(getBaseContext(), R.color.colorWhite), android.graphics.PorterDuff.Mode.SRC_IN);
    }

    @Override
    protected void onPostResume() {
        super.onPostResume();
        onInitView();
    }

    private void onInitView() {
        mUserInput = "";
        for (int i = 0; i < mIvCircle.length; i++) {
            mIvCircle[i].setBackground(getDrawable(R.drawable.ic_pass_gr));
        }
        mViewPager.setCurrentItem(0, true);
        onCheckFingerPrint();
    }

    @Override
    public void onBackPressed() {
        if (mCancellationSignal != null)
            mCancellationSignal.cancel();
        moveTaskToBack(true);
    }

    private void onUnlock() {
        if (mCancellationSignal != null)
            mCancellationSignal.cancel();
        finish();
        overridePendingTransition(R.anim.fade_in, R.anim.slide_out_bottom);
    }

    @TargetApi(Build.VERSION_CODES.M)
    private void onCheckFingerPrint() {
        mFingerprintManagerCompat = FingerprintManagerCompat.from(this);
        mCancellationSignal = new CancellationSignal();
        if ((Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) &&
                mFingerprintManagerCompat.isHardwareDetected() &&
                mFingerprintManagerCompat.hasEnrolledFingerprints() &&
                getBaseDao().getUsingFingerPrint()) {
            mFingerImage.setVisibility(View.VISIBLE);
            mUnlockMsg.setText(R.string.str_app_unlock_msg2);

            mFingerprintManagerCompat.authenticate(null, 0, mCancellationSignal, new FingerprintManagerCompat.AuthenticationCallback() {
                @Override
                public void onAuthenticationError(int errMsgId, CharSequence errString) {
                    super.onAuthenticationError(errMsgId, errString);
                }

                @Override
                public void onAuthenticationHelp(int helpMsgId, CharSequence helpString) {
                    mUnlockMsg.setText(helpString);
                    mFingerImage.setColorFilter(ContextCompat.getColor(getBaseContext(), R.color.colorRed), android.graphics.PorterDuff.Mode.SRC_IN);
                    super.onAuthenticationHelp(helpMsgId, helpString);
                }

                @Override
                public void onAuthenticationSucceeded(FingerprintManagerCompat.AuthenticationResult result) {
                    super.onAuthenticationSucceeded(result);
                    mFingerImage.setColorFilter(ContextCompat.getColor(getBaseContext(), R.color.colorAtom), android.graphics.PorterDuff.Mode.SRC_IN);
                    onUnlock();
                }

                @Override
                public void onAuthenticationFailed() {
                    super.onAuthenticationFailed();
                }


            }, null);

        } else {
            mFingerImage.setVisibility(View.GONE);
            mUnlockMsg.setText(R.string.str_app_unlock_msg1);
        }
    }

    @Override
    public void userInsertKey(char input) {
        if (mUserInput == null || mUserInput.length() == 0) {
            mUserInput = String.valueOf(input);

        } else if (mUserInput.length() < 5) {
            mUserInput = mUserInput + input;
        }

        if (mUserInput.length() == 4) {
            mViewPager.setCurrentItem(1, true);

        } else if (mUserInput.length() == 5 && WUtil.checkPasscodePattern(mUserInput)) {
            onFinishInput();

        } else if (mUserInput.length() == 5 && !WUtil.checkPasscodePattern(mUserInput)) {
            onInitView();
            return;
        }
        onUpdateCnt();
    }

    @Override
    public void userDeleteKey() {
        if (mUserInput == null || mUserInput.length() <= 0) {
            onBackPressed();
        } else if (mUserInput.length() == 4) {
            mUserInput = mUserInput.substring(0, mUserInput.length() - 1);
            mViewPager.setCurrentItem(0, true);
        } else {
            mUserInput = mUserInput.substring(0, mUserInput.length() - 1);
        }
        onUpdateCnt();
    }

    private void onFinishInput() {
        onShowWaitDialog();
        new CheckPasswordTask(getBaseApplication(), this).execute(mUserInput);
    }

    private void onUpdateCnt() {
        if (mUserInput == null)
            mUserInput = "";

        final int inputLength = mUserInput.length();
        for (int i = 0; i < mIvCircle.length; i++) {
            if (i < inputLength)
                mIvCircle[i].setBackground(ContextCompat.getDrawable(AppLockActivity.this, R.drawable.ic_pass_pu));
            else
                mIvCircle[i].setBackground(ContextCompat.getDrawable(AppLockActivity.this, R.drawable.ic_pass_gr));
        }
    }

    private void onShakeView() {
        mLayerContents.clearAnimation();
        Animation animation = AnimationUtils.loadAnimation(this, R.anim.shake);
        animation.reset();
        animation.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {
            }

            @Override
            public void onAnimationEnd(Animation animation) {
                onInitView();
            }

            @Override
            public void onAnimationRepeat(Animation animation) {
            }
        });
        mLayerContents.startAnimation(animation);
    }


    @Override
    protected void onStop() {
        super.onStop();
        if (mCancellationSignal != null)
            mCancellationSignal.cancel();
    }

    @Override
    public void onTaskResponse(TaskResult result) {
        if (isFinishing()) return;
        onHideWaitDialog();
        if (result.isSuccess) {
            onUnlock();
        } else {
            onShakeView();
            onInitView();
            Toast.makeText(getBaseContext(), getString(R.string.error_invalid_password), Toast.LENGTH_SHORT).show();
        }
    }


    public class KeyboardPagerAdapter extends FragmentPagerAdapter {

        private ArrayList<KeyboardFragment> mFragments = new ArrayList<>();

        public KeyboardPagerAdapter(FragmentManager fm) {
            super(fm);
            mFragments.clear();
            NumberKeyBoardFragment number = NumberKeyBoardFragment.newInstance();
            number.setListener(AppLockActivity.this);
            mFragments.add(number);

            AlphabetKeyBoardFragment alphabet = AlphabetKeyBoardFragment.newInstance();
            alphabet.setListener(AppLockActivity.this);
            mFragments.add(alphabet);
        }

        @Override
        public Fragment getItem(int position) {
            return mFragments.get(position);
        }

        @Override
        public int getCount() {
            return mFragments.size();
        }

        public ArrayList<KeyboardFragment> getFragments() {
            return mFragments;
        }
    }
}

File: app/src/main/java/wannabit/io/cosmostaion/ui/password/AppLockActivity.kt
package wannabit.io.cosmostaion.ui.password

import android.os.Bundle
import android.view.WindowManager
import android.widget.ImageView
import androidx.core.hardware.fingerprint.FingerprintManagerCompat
import androidx.core.os.CancellationSignal
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.ViewModelProvider
import androidx.viewpager2.adapter.FragmentStateAdapter
import wannabit.io.cosmostaion.R
import wannabit.io.cosmostaion.common.BaseActivity
import wannabit.io.cosmostaion.common.BaseConstant
import wannabit.io.cosmostaion.common.BaseData
import wannabit.io.cosmostaion.common.BaseUtils.checkPasscodePattern
import wannabit.io.cosmostaion.common.makeToast
import wannabit.io.cosmostaion.data.repository.wallet.WalletRepositoryImpl
import wannabit.io.cosmostaion.database.Prefs
import wannabit.io.cosmostaion.databinding.ActivityPasswordCheckBinding
import wannabit.io.cosmostaion.ui.viewmodel.intro.WalletViewModel
import wannabit.io.cosmostaion.ui.viewmodel.intro.WalletViewModelProviderFactory

class AppLockActivity : BaseActivity(), KeyboardListener {

    private lateinit var binding: ActivityPasswordCheckBinding

    private lateinit var passwordPageAdapter: PasswordPageAdapter

    private lateinit var walletViewModel: WalletViewModel

    private val ivCircle = arrayOfNulls<ImageView>(5)
    private var userInput = ""

    private val cancellationSignal: CancellationSignal? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityPasswordCheckBinding.inflate(layoutInflater)
        setContentView(binding.root)

        initViewModel()
        initView()
        onUpdateView()
        checkPwObserve()

        if (Prefs.usingBio) {
            checkFingerPrint()
        }
    }

    private fun initViewModel() {
        val walletRepository = WalletRepositoryImpl()
        val walletViewModelProviderFactory = WalletViewModelProviderFactory(walletRepository)
        walletViewModel = ViewModelProvider(this, walletViewModelProviderFactory)[WalletViewModel::class.java]
    }

    private fun initView() {
        window.setFlags(
            WindowManager.LayoutParams.FLAG_SECURE,
            WindowManager.LayoutParams.FLAG_SECURE
        )
        for (i in ivCircle.indices) {
            ivCircle[i] = findViewById(resources.getIdentifier("pin_code_img$i", "id", packageName))
        }

        passwordPageAdapter = PasswordPageAdapter(this)
        binding.apply {
            pagerKeyboard.adapter = passwordPageAdapter
            pagerKeyboard.isUserInputEnabled = false
            pagerKeyboard.offscreenPageLimit = 2
        }
    }

    private fun onUpdateView() {
        userInput = ""
        for (i in ivCircle.indices) {
            ivCircle[i]?.setImageResource(R.drawable.icon_pin_code_disable)
        }
        binding.pagerKeyboard.setCurrentItem(0, true)
    }

    override fun userInsertPassword(input: Char) {
        if (userInput.isEmpty()) {
            userInput = input.toString()
        } else if (userInput.length < 5) {
            userInput += input
        }

        if (userInput.length == 4) {
            binding.pagerKeyboard.setCurrentItem(1, true)
        } else if (userInput.length == 5 && checkPasscodePattern(userInput)) {
            walletViewModel.checkPassword(userInput)
        } else if (userInput.length == 5 && !checkPasscodePattern(userInput)) {
            onUpdateView()
            return
        }
        updatePin()
    }

    override fun userDeletePassword() {
        if (userInput.isEmpty()) {
            onBackPressed()
        } else if (userInput.length == 4) {
            userInput = userInput.substring(0, userInput.length - 1)
            binding.pagerKeyboard.setCurrentItem(0, true)
        } else {
            userInput = userInput.substring(0, userInput.length - 1)
        }
        updatePin()
    }

    private fun updatePin() {
        if (userInput.isEmpty()) userInput = ""

        val inputLength: Int = userInput.length
        for (i in ivCircle.indices) {
            if (i < inputLength) ivCircle[i]?.setImageResource(R.drawable.icon_pin_code_active)
            else ivCircle[i]?.setImageResource(R.drawable.icon_pin_code_disable)
        }
    }

    private fun checkPwObserve() {
        walletViewModel.pwCheckResult.observe(this) { result ->
            if (result == BaseConstant.SUCCESS) {
                BaseData.setLastTime()
                cancellationSignal?.cancel()
                finish()
                overridePendingTransition(R.anim.anim_fade_in, R.anim.anim_slide_out_bottom)

            } else {
                onUpdateView()
                makeToast(R.string.error_invalid_password)
            }
        }
    }

    private fun checkFingerPrint() {
        val fingerprintManagerCompat = FingerprintManagerCompat.from(this)
        val cancellationSignal = CancellationSignal()
        if (fingerprintManagerCompat.isHardwareDetected && fingerprintManagerCompat.hasEnrolledFingerprints() && Prefs.usingBio) {
            fingerprintManagerCompat.authenticate(
                null,
                0,
                cancellationSignal,
                object : FingerprintManagerCompat.AuthenticationCallback() {
                    override fun onAuthenticationError(errMsgId: Int, errString: CharSequence) {
                        super.onAuthenticationError(errMsgId, errString)
                    }

                    override fun onAuthenticationHelp(helpMsgId: Int, helpString: CharSequence) {
                        super.onAuthenticationHelp(helpMsgId, helpString)
                    }

                    override fun onAuthenticationSucceeded(result: FingerprintManagerCompat.AuthenticationResult) {
                        super.onAuthenticationSucceeded(result)
                        cancellationSignal.cancel()
                        setResult(RESULT_OK, intent)
                        finish()
                        overridePendingTransition(R.anim.anim_fade_in, R.anim.anim_slide_out_bottom)
                    }

                    override fun onAuthenticationFailed() {
                        super.onAuthenticationFailed()
                    }
                },
                null
            )
        }
    }

    override fun onBackPressed() {
        cancellationSignal?.cancel()
        moveTaskToBack(true)
    }

    override fun onStop() {
        super.onStop()
        cancellationSignal?.cancel()
    }

    class PasswordPageAdapter(fragmentActivity: FragmentActivity) : FragmentStateAdapter(fragmentActivity) {

        private val fragmentList = mutableListOf<KeyboardFragment>()

        init {
            fragmentList.clear()
            val number = NumberKeyBoardFragment()
            number.setKeyboardListener(fragmentActivity as AppLockActivity)
            fragmentList.add(number)

            val alphabet = AlphabetKeyBoardFragment()
            alphabet.setKeyboardListener(fragmentActivity)
            fragmentList.add(alphabet)
        }

        override fun getItemCount(): Int {
            return fragmentList.size
        }

        override fun createFragment(position: Int): Fragment {
            return fragmentList[position]
        }
    }
}